app/node/distort/cornerpin/cornerpindistortnode.cpp

```diff
@@ -0,0 +1,237 @@
+/***
+
+  Olive - Non-Linear Video Editor
+  Copyright (C) 2021 Olive Team
+
+  This program is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+***/
+
+#include "cornerpindistortnode.h"
+
+#include "common/lerp.h"
+#include "core.h"
+#include "widget/slider/floatslider.h"
+
+namespace olive {
+
+const QString CornerPinDistortNode::kTextureInput = QStringLiteral("tex_in");
+const QString CornerPinDistortNode::kTopLeftInput = QStringLiteral("top_left_in");
+const QString CornerPinDistortNode::kTopRightInput = QStringLiteral("top_right_in");
+const QString CornerPinDistortNode::kBottomRightInput = QStringLiteral("bottom_right_in");
+const QString CornerPinDistortNode::kBottomLeftInput = QStringLiteral("bottom_left_in");
+const QString CornerPinDistortNode::kPerspectiveInput = QStringLiteral("perspective_in");
+
+CornerPinDistortNode::CornerPinDistortNode()
+{
+  AddInput(kTextureInput, NodeValue::kTexture, InputFlags(kInputFlagNotKeyframable));
+  AddInput(kPerspectiveInput, NodeValue::kBoolean, true);
+  AddInput(kTopLeftInput, NodeValue::kVec2, QVector2D(0.0, 0.0));
+  AddInput(kTopRightInput, NodeValue::kVec2, QVector2D(0.0, 0.0));
+  AddInput(kBottomRightInput, NodeValue::kVec2, QVector2D(0.0, 0.0));
+  AddInput(kBottomLeftInput, NodeValue::kVec2, QVector2D(0.0, 0.0));
+}
+
+void CornerPinDistortNode::Retranslate()
+{
+  SetInputName(kTextureInput, tr("Texture"));
+  SetInputName(kPerspectiveInput, tr("Perspective"));
+  SetInputName(kTopLeftInput, tr("Top Left"));
+  SetInputName(kTopRightInput, tr("Top Right"));
+  SetInputName(kBottomRightInput, tr("Bottom Right"));
+  SetInputName(kBottomLeftInput, tr("Bottom Left"));
+}
+
+void CornerPinDistortNode::Value(const NodeValueRow &value, const NodeGlobals &globals, NodeValueTable *table) const
+{
+  ShaderJob job;
+  job.InsertValue(value);
+  job.SetAlphaChannelRequired(GenerateJob::kAlphaForceOn);
+  job.InsertValue(QStringLiteral("resolution_in"), NodeValue(NodeValue::kVec2, globals.resolution(), this));
+
+  // Convert slider values to their pixel values and then convert to clip space (-1.0 ... 1.0) for overriding the
+  // vertex coordinates.
+  const QVector2D &resolution = globals.resolution();
+  QVector2D top_left = QVector2D(ValueToPixel(0, value, resolution)) / (resolution / 2.0) - QVector2D(1.0, 1.0);
```

Comment:
Might be good to cache `resolution / 2.0` instead of calculating it each time, also probably preferable to multiply by 0.5 instead (multiplication is generally faster than division)